Turn into installable package

fsimonis opened the pull request:
This PR turns the `ccx2paraview` script into an installable package by

1. creating a python package `ccx2paraview`
2. splitting the current script into functionality `common.py` and interface `cli.y`
3. creating a `pyproject.toml` and a stub `setup.py`

I also did some cleanup of the repo.

What this achieves:

* Publishes the project description in `pyproject.toml` to PyPi.
* Allows to directly install and use from PyPi `pipx install ccx2paraview` or via a venv and `pip install ccx2paraview` then `ccx2paraview case.frd vtk`
* Can be imported `from ccx2paraview.common import Converter` for integration with other tools
* Simplifies additions, like aliases such as `frd2vtk`
* Automatically uses version from the git version tag

Todos:

* [ ] Iterate over documentation if changes are in-line with reviewer
* [ ] setup a workflow that automatically publishes a new PyPi version whenever a new tag is pushed. [Example workflow](https://github.com/precice/fmi-runner/blob/main/.github/workflows/pythonpublish.yml). This requires `PYPI_TOKEN` to be setup as a repository secret. You can get that from the PiPy account.

If you want to try the installation directly from my fork:
```
pipx install https://github.com/fsimonis/ccx2paraview/archive/refs/heads/add-pyproject.zip
```

fsimonis committed:
Turn script into package

fsimonis committed:
Cleanup scripts

fsimonis committed:
Add project definition files

fsimonis committed:
Add ignores for dist build

fsimonis committed:
Remove obsolete requirements.txt

fsimonis commented on pyproject.toml:
@imirzov Could you check if the information in this file is correct?

imirzov replied:
Yep, almost. I would edit keywords and description according to the original repo.

fsimonis commented on pyproject.toml:
```diff
@@ -0,0 +1,40 @@
+"Bug Tracker" = "https://github.com/calculix/ccx2paraview/issues"
+
+[project.scripts]
+ccx2paraview = "ccx2paraview.cli:main"
```
FYI this defining an executable script.

`ccx2paraview` is a script that calls function `main` from the package `ccs2paraview.cli`.
This is very easy to extend with convenience scripts.

imirzov commented:
Hi. 
ccx2paraview is already an installable package and is present on pypi.org: https://pypi.org/project/ccx2paraview/
Command "pip install ccx2paraview" works fine.
The converter already could be used within other codes - see section "For developers" in the README.md.
So, looks like you provided no new functionality for CalculiX users.

You have removed requirements.txt, but it's needed.
You have split 1 code file into 2. I don't like it.

Sorry, I will not merge the PR until reasoned explanations about WTF are you doing.

fsimonis commented:
Hi,

The reason behind this entire PR is that I would like to install and use your script like an application from both PyPi and a local clone of the repository:

```console
pipx install ccx2paraview
ccx2paraview yourjobname.frd vtk
```

This is currently not possible with the package on PyPi, due to the way it is packaged.

The package does not contain metadata regarding dependencies. Which makes isolated installations non-function (something pipx relies on):

```console
$ python -m venv venv && . venv/bin/activate
$ pip install ccx2paraview
$ python -m ccx2paraview.ccx2paraview
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/.../venv/lib/python3.11/site-packages/ccx2paraview/ccx2paraview.py", line 23, in <module>
    import numpy as np
ModuleNotFoundError: No module named 'numpy'
```

To add this dependency information, I added the [`pyproject.toml`](https://packaging.python.org/en/latest/guides/writing-pyproject-toml/). This replaces the `requirements.txt`. (It's last use is to pin dependency versions for application deployment, which isn't used here.) This `pyproject.toml` has the added bonus, that you can control the entire metadata shown on PyPi directly from within the repo.

Next step was to define [a console entry-point](https://packaging.python.org/en/latest/specifications/entry-points/) for your python module to make it usable like an executable. This required the name of the entry-point and the function to call in a `entrypoint = package:function` notation. In practise, an entrypoint is an executable wrapper that imports the package and calls the given function.

As your repository currently only provides the python module, I needed to restructure it, moving it into a package, hence the `ccx2paraview` subdirectory.

Specifying the double `ccx2paraview.ccx2paraview` looked a bit silly. As far as I can tell, the common way to refactor this is to split the module into implementation + CLI, which is what I did.

So, I moved the CLI to `ccx2paraview/cli.py` and the implementation to `ccx2paraview/common.cli`, which allowed to specify the entrypoint as:

```toml
[project.scripts]
ccx2paraview = "ccx2paraview.cli:main"
```

Of course splitting the script comes at the downside that it cannot simply be copied to a machine with restricted internet access.

The added functionality for CalculiX users is that they can now install and use your python module way easier either.
Hope that explains the motivation behind this PR.

imirzov commented:
Ok, merged!

Kindly ask you to actualize README.txt and pyproject.toml. Correct keywords are `converter vtk vtu paraview calculix frd`, correct description is `CalculiX to Paraview converter (frd to vtk/vtu). Makes possible to view and postprocess CalculiX analysis results in Paraview. Generates Mises and Principal components for stress and strain tensors.`

Next you can draft a new release v3.2.0 and publish it on PyPi.

Thanks for your contribution! Please, do not stop and continue.
